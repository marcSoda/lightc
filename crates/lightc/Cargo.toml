[package]
name = "lightc"
version = "0.1.0"
edition = "2021"

[dependencies]
codegen = { path = "../codegen" }
common = { path = "../common" }
lexer = { path = "../lexer" }
parser = { path = "../parser" }
type_checker = { path = "../type_checker" }

clap = { version = "3.0.14", features = ["derive"] }
inkwell = { git = "https://github.com/TheDan64/inkwell", branch = "master", features = ["llvm13-0"] }
llvm-sys = "130"
tempfile = "3.3.0"
