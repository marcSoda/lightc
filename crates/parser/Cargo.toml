[package]
name = "parser"
version = "0.0.0"
edition = "2021"

[dependencies]
serde = { version = "1.0.136", features = ["derive"] }

ast = { path = "../ast" }
common = { path = "../common" }
symbol_table = { path = "../symbol_table" }

[dev-dependencies]
insta = { version = "1.12.0", features = ["glob"] }

lexer = { path = "../lexer" }
