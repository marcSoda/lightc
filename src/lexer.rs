use std::iter::Peekable;
use serde::Serialize;

use crate::token::{Symbol, Token, TokenType};

type LexResult = std::result::Result<Token, LexError>;

pub(crate) struct Lexer {
    stream: Peekable<StreamIter<char>>,
}

impl Lexer {
    pub fn new(input: &str) -> Self {
        Lexer {
            stream: StreamIter::new(input).peekable(),
        }
    }

    fn lex(&mut self) -> LexResult {
        let cur = match self.stream.next() {
            Some(cur) => cur,
            None => return Ok(Token::default()),
        };

        if cur.value.is_ascii_whitespace() {
            while let Some(c) = self.stream.peek() {
                if !c.value.is_ascii_whitespace() {
                    return self.lex();
                }
                self.stream.next();
            }
            return self.lex(); // Eat trailing newline
        }

        // Comments
        if cur == '/' && matches!(self.stream.peek(), Some(c) if *c == '/') {
            while let Some(c) = self.stream.next() {
                if c == '\n' {
                    return self.lex();
                }
            }
            return self.lex(); // Eat trailing comment
        }

        // Identifiers: keywords, function names, and variables
        if cur.value.is_ascii_alphabetic() {
            let mut identifier = String::from(cur.value);
            while let Some(c) = self.stream.peek() {
                if c.value.is_ascii_alphanumeric() || *c == '_' {
                    identifier.push(c.value);
                    self.stream.next();
                } else {
                    break;
                }
            }

            let tt = match identifier.as_str() {
                "else" => TokenType::Else,
                "extern" => TokenType::Extern,
                //"u64" => TokenTypeVarType(TokenTypeU64),
                "fn" => TokenType::Fn,
                "for" => TokenType::For,
                "if" => TokenType::If,
                "let" => TokenType::Let,
                _ => TokenType::Ident(identifier),
            };

            return Ok(Token::from((tt, cur)));
        }

        // Numbers
        if cur.value.is_ascii_digit() {
            let mut num = String::from(cur.value);
            while let Some(c) = self.stream.peek() {
                if c.value.is_ascii_alphanumeric() || *c == '.' {
                    num.push(c.value);
                    self.stream.next();
                } else {
                    break;
                }
            }

            return match num.parse() {
                Ok(n) => Ok(Token::from((TokenType::Int(n), cur))),
                Err(_) => Err(LexError::from((format!("Invalid number: {}", num), cur))),
            };
        }

        // Logical operators
        if let Some(next) = self.stream.peek() {
            match cur.value {
                '=' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::from((TokenType::Op(Symbol::Eq), cur)));
                }
                '&' if next == &'&' => {
                    self.stream.next();
                    return Ok(Token::from((TokenType::Op(Symbol::And), cur)));
                }
                '|' if next == &'|' => {
                    self.stream.next();
                    return Ok(Token::from((TokenType::Op(Symbol::Or), cur)));
                }
                _ => (),
            }
        }

        // Everything else
        let tt = match cur.value {
            '+' => TokenType::Op(Symbol::Plus),
            '-' => TokenType::Op(Symbol::Minus),
            '*' => TokenType::Op(Symbol::Mult),
            '/' => TokenType::Op(Symbol::Div),
            '^' => TokenType::Op(Symbol::Pow),
            '>' => TokenType::Op(Symbol::Gt),
            '<' => TokenType::Op(Symbol::Lt),
            '!' => TokenType::Op(Symbol::Not),
            '=' => TokenType::Op(Symbol::Assign),
            '}' => TokenType::CloseBrace,
            ')' => TokenType::CloseParen,
            ':' => TokenType::Colon,
            ',' => TokenType::Comma,
            '{' => TokenType::OpenBrace,
            '(' => TokenType::OpenParen,
            ';' => TokenType::Semicolon,
            c => {
                return Err(LexError::from((format!("Unknown character: {}", c), cur)));
            }
        };

        Ok(Token::from((tt, cur)))
    }
}

impl Iterator for Lexer {
    type Item = LexResult;

    fn next(&mut self) -> Option<Self::Item> {
        match self.lex() {
            Ok(Token {
                tt: TokenType::Eof, ..
            }) => None,
            x => Some(x),
        }
    }
}

// Provides additional context for each source character
#[derive(Debug, PartialEq)]
struct ContextPoint<T> {
    value: T,
    line: usize,
    column: usize,
}

impl<T> ContextPoint<T> {
    fn new(value: T, line: usize, column: usize) -> Self {
        ContextPoint {
            value,
            line: line + 1,
            column: column + 1,
        }
    }
}

impl PartialEq<char> for ContextPoint<char> {
    fn eq(&self, other: &char) -> bool {
        self.value == *other
    }
}

impl<T> From<(TokenType, ContextPoint<T>)> for Token {
    fn from((ty, ContextPoint { line, column, .. }): (TokenType, ContextPoint<T>)) -> Self {
        Token {
            tt: ty,
            line,
            column,
        }
    }
}

// Iterator for the source character stream. Supports line and column context.
struct StreamIter<T> {
    lines: Vec<Vec<T>>,
    line: usize,
    column: usize,
}

impl StreamIter<char> {
    fn new(input: &str) -> Self {
        StreamIter {
            lines: input
                .split_inclusive('\n') // can't use .lines() or we lose '\n'
                .map(|l| l.chars().collect())
                .collect(),
            line: 0,
            column: 0,
        }
    }
}

impl Iterator for StreamIter<char> {
    type Item = ContextPoint<char>;

    fn next(&mut self) -> Option<Self::Item> {
        let line = self.lines.get(self.line);
        let cc = line?
            .get(self.column)
            .map(|c| ContextPoint::new(*c, self.line, self.column))
            .or_else(|| {
                self.line += 1;
                self.column = 0;
                self.lines.get(self.line).and_then(|line| {
                    line.get(self.column)
                        .map(|c| ContextPoint::new(*c, self.line, self.column))
                })
            });
        self.column += 1;
        cc
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub(crate) struct LexError {
    message: String,
    line: usize,
    column: usize,
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Lexing error: {} at {}:{}",
            self.message, self.line, self.column
        )
    }
}

impl std::error::Error for LexError {}

impl<T> From<(String, ContextPoint<T>)> for LexError {
    fn from((msg, cp): (String, ContextPoint<T>)) -> Self {
        LexError {
            message: msg,
            line: cp.line,
            column: cp.column,
        }
    }
}

#[cfg(test)]
mod test {
    use insta::{assert_yaml_snapshot, glob, with_settings};
    use std::fs;

    use super::*;

    #[test]
    fn test_lexer() {
        with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            glob!("tests/inputs/lexer/*.input", |path| {
                let input = fs::read_to_string(path).unwrap();
                assert_yaml_snapshot!(Lexer::new(&input).collect::<Result<Vec<_>, _>>());
            });
        });
    }

    #[test]
    fn test_stream_next() {
        let input = "\
abc
def
ghi
";
        let mut stream = StreamIter::new(input);
        assert_eq!(ContextPoint::new('a', 0, 0), stream.next().unwrap());
        assert_eq!(ContextPoint::new('b', 0, 1), stream.next().unwrap());
        assert_eq!(ContextPoint::new('c', 0, 2), stream.next().unwrap());
        assert_eq!(ContextPoint::new('\n', 0, 3), stream.next().unwrap());
        assert_eq!(ContextPoint::new('d', 1, 0), stream.next().unwrap());
        assert_eq!(ContextPoint::new('e', 1, 1), stream.next().unwrap());
        assert_eq!(ContextPoint::new('f', 1, 2), stream.next().unwrap());
        assert_eq!(ContextPoint::new('\n', 1, 3), stream.next().unwrap());
        assert_eq!(ContextPoint::new('g', 2, 0), stream.next().unwrap());
        assert_eq!(ContextPoint::new('h', 2, 1), stream.next().unwrap());
        assert_eq!(ContextPoint::new('i', 2, 2), stream.next().unwrap());
        assert_eq!(ContextPoint::new('\n', 2, 3), stream.next().unwrap());
        assert_eq!(None, stream.next());
        assert_eq!(None, stream.next());
    }

    #[test]
    fn test_lex() {
        use Symbol::*;
        use TokenType::*;

        let input = "\
let foo = 14
// bar
/ not_a_comment
baz
";
        let mut lexer = Lexer::new(input);
        assert_eq!(Ok(Token::new(Let, 1, 1)), lexer.lex());
        assert_eq!(Ok(Token::new(Ident("foo".to_string()), 1, 5)), lexer.lex());
        assert_eq!(Ok(Token::new(Op(Assign), 1, 9)), lexer.lex());
        assert_eq!(Ok(Token::new(Int(14), 1, 11)), lexer.lex());
        assert_eq!(Ok(Token::new(Op(Div), 3, 1)), lexer.lex());
        assert_eq!(
            Ok(Token::new(Ident("not_a_comment".to_string()), 3, 3)),
            lexer.lex()
        );
        assert_eq!(Ok(Token::new(Ident("baz".to_string()), 4, 1)), lexer.lex());
        assert_eq!(Ok(Token::default()), lexer.lex());
    }
}
