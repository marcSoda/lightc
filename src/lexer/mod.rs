use serde::Serialize;
use std::iter::Peekable;

use crate::token::{Symbol, Token, TokenType, Type};

type LexResult = std::result::Result<Token, LexError>;

pub(crate) struct Lexer {
    stream: Peekable<StreamIter<char>>,
    pub tokens: Vec<Token>,
}

impl Lexer {
    pub fn new(input: &str) -> Self {
        Lexer {
            stream: StreamIter::new(input).peekable(),
            tokens: vec![],
        }
    }

    // Scan all input
    pub fn scan(mut self) -> Result<Vec<Token>, LexError> {
        loop {
            let token = self.lex()?;
            if token.is_eof() {
                break;
            }
            self.tokens.push(token);
        }
        Ok(self.tokens)
    }

    // Recursively process enough characters to produce one token
    fn lex(&mut self) -> LexResult {
        let cur = match self.stream.next() {
            Some(cur) => cur,
            None => unreachable!("fatal: can't lex nothing"),
        };

        // Inject a semicolon if certain tokens occur at the end of the line or
        // EOF. If EOF, make sure the context is right.
        if cur.value == '\n' && self.should_add_semicolon() {
            return Ok(Token::from((TokenType::Semicolon(true), cur)));
        } else if cur.is_eof() {
            if self.should_add_semicolon() {
                let semi = match self.tokens.last() {
                    Some(t) => Token::new(TokenType::Semicolon(true), t.line, t.column + 1),
                    None => Token::default(),
                };
                self.tokens.push(semi);
            }
            return Ok(Token::from((TokenType::Eof, cur)));
        }

        // Ignore whitespace
        if cur.value.is_ascii_whitespace() {
            while let Some(c) = self.stream.peek() {
                if !c.value.is_ascii_whitespace() {
                    return self.lex();
                } else if c.is_eof() {
                    break;
                }
                self.stream.next();
            }
            return self.lex(); // Eat trailing newline
        }

        // Single line comments
        if cur == '/' && matches!(self.stream.peek(), Some(c) if *c == '/') {
            while let Some(c) = self.stream.next() {
                if c == '\n' {
                    return self.lex();
                } else if c.is_eof() {
                    break;
                }
            }
            return self.lex(); // Eat trailing comment
        }

        // Keywords, types, and identifiers
        if cur.value.is_ascii_alphabetic() {
            let mut identifier = String::from(cur.value);
            while let Some(c) = self.stream.peek() {
                if c.value.is_ascii_alphanumeric() || *c == '_' {
                    identifier.push(c.value);
                    self.stream.next();
                } else {
                    break;
                }
            }

            let tt = match identifier.as_str() {
                "fn" => TokenType::Fn,
                "let" => TokenType::Let,
                "for" => TokenType::For,
                "if" => TokenType::If,
                "else" => TokenType::Else,
                "extern" => TokenType::Extern,
                "true" => TokenType::Bool(true),
                "false" => TokenType::Bool(false),
                "int8" => TokenType::VarType(Type::Int8),
                "int16" => TokenType::VarType(Type::Int16),
                "int32" => TokenType::VarType(Type::Int32),
                "int64" => TokenType::VarType(Type::Int64),
                "uint8" => TokenType::VarType(Type::UInt8),
                "uint16" => TokenType::VarType(Type::UInt16),
                "uint32" => TokenType::VarType(Type::UInt32),
                "uint64" => TokenType::VarType(Type::UInt64),
                "float" => TokenType::VarType(Type::Float),
                "double" => TokenType::VarType(Type::Double),
                "bool" => TokenType::VarType(Type::Bool),
                "char" => TokenType::VarType(Type::Char),
                // TODO: don't hardcode these
                "int" => TokenType::VarType(Type::Int32),
                "uint" => TokenType::VarType(Type::UInt32),
                _ => TokenType::Ident(identifier),
            };

            return Ok(Token::from((tt, cur)));
        }

        // Literal numbers
        if cur.value.is_ascii_digit() {
            let mut n = String::from(cur.value);
            while let Some(c) = self.stream.peek() {
                if c.value.is_ascii_alphanumeric() || *c == '.' {
                    n.push(c.value);
                    self.stream.next();
                } else {
                    break;
                }
            }

            return Ok(Token::from((TokenType::Num(n), cur)));
        }

        // Literal char
        if cur == '\'' {
            let mut ch = String::new();
            let next = self
                .stream
                .next()
                .unwrap_or_else(|| unreachable!("fatal: lexed None when looking for char"));

            match next.value {
                // Control characters
                '\\' => {
                    if let Some(next) = self.stream.next() {
                        match next.value {
                            'n' => ch = String::from("\n"),
                            't' => ch = String::from("\t"),
                            '\'' => ch = String::from("'"),
                            c => {
                                return Err(LexError::from((
                                    format!("Invalid character control sequence: `\\{}`", c),
                                    next,
                                )))
                            }
                        }
                    }
                }
                // EOF
                '\0' => {
                    return Err(LexError::from((
                        "Unterminated character literal. Expecting `'`, got `EOF`".to_string(),
                        cur,
                    )));
                }
                '\'' => {
                    return Err(LexError::from((
                        "Character literal can't be empty".to_string(),
                        cur,
                    )))
                }

                // Everything else
                c => ch = String::from(c),
            }

            // Check for closing '\''
            let last = self
                .stream
                .next()
                .unwrap_or_else(|| unreachable!("fatal: lexed None when looking for `'`"));
            match last.value {
                '\'' => (),
                '\0' | '\n' => {
                    return Err(LexError::from((
                        "Unterminated character literal. Expecting `'`".to_string(),
                        last,
                    )));
                }
                _ => {
                    return Err(LexError::from((
                        format!("Invalid character sequence: `'{}{}'`", ch, last.value),
                        last,
                    )));
                }
            }

            return Ok(Token::from((TokenType::Char(ch), cur)));
        }

        // Multi-character operators
        if let Some(next) = self.stream.peek() {
            match cur.value {
                '=' if next == &'=' => {
                    self.stream.next();
                    return Ok(Token::from((TokenType::Op(Symbol::Eq), cur)));
                }
                '&' if next == &'&' => {
                    self.stream.next();
                    return Ok(Token::from((TokenType::Op(Symbol::And), cur)));
                }
                '|' if next == &'|' => {
                    self.stream.next();
                    return Ok(Token::from((TokenType::Op(Symbol::Or), cur)));
                }
                '-' if next == &'>' => {
                    self.stream.next();
                    return Ok(Token::from((TokenType::Op(Symbol::RetType), cur)));
                }
                _ => (),
            }
        }

        // Everything else
        let tt = match cur.value {
            '+' => TokenType::Op(Symbol::Add),
            '-' => TokenType::Op(Symbol::Sub),
            '*' => TokenType::Op(Symbol::Mul),
            '/' => TokenType::Op(Symbol::Div),
            '^' => TokenType::Op(Symbol::Pow),
            '>' => TokenType::Op(Symbol::Gt),
            '<' => TokenType::Op(Symbol::Lt),
            '!' => TokenType::Op(Symbol::Not),
            '=' => TokenType::Op(Symbol::Assign),
            '}' => TokenType::CloseBrace,
            ')' => TokenType::CloseParen,
            ':' => TokenType::Colon,
            ',' => TokenType::Comma,
            '{' => TokenType::OpenBrace,
            '(' => TokenType::OpenParen,
            ';' => TokenType::Semicolon(false),
            c => {
                return Err(LexError::from((format!("Unknown character: {}", c), cur)));
            }
        };

        Ok(Token::from((tt, cur)))
    }

    // Add a semicolon for these tokens
    fn should_add_semicolon(&self) -> bool {
        use TokenType::*;

        if let Some(t) = self.tokens.last() {
            matches!(
                t.tt,
                Bool(_) | Char(_) | CloseBrace | CloseParen | Ident(_) | Num(_) | VarType(_)
            )
        } else {
            false
        }
    }
}

// Provides additional context for each source character
#[derive(Debug, Clone, Copy, PartialEq)]
struct ContextElement<T> {
    value: T,
    line: usize,
    column: usize,
}

impl<T> ContextElement<T> {
    fn new(value: T, line: usize, column: usize) -> Self {
        ContextElement {
            value,
            line: line + 1,
            column: column + 1,
        }
    }
}

impl ContextElement<char> {
    // Will cause lexing to stop if there's a null byte in the file
    fn is_eof(&self) -> bool {
        self.value == 0 as char
    }
}

impl PartialEq<char> for ContextElement<char> {
    fn eq(&self, other: &char) -> bool {
        self.value == *other
    }
}

impl<T> From<(TokenType, ContextElement<T>)> for Token {
    fn from((ty, ContextElement { line, column, .. }): (TokenType, ContextElement<T>)) -> Self {
        Token {
            tt: ty,
            line,
            column,
        }
    }
}

// Iterator for the source character stream. Supports line and column context.
struct StreamIter<T> {
    lines: Vec<Vec<T>>,
    line: usize,
    column: usize,
}

impl StreamIter<char> {
    fn new(input: &str) -> Self {
        StreamIter {
            lines: input
                .split_inclusive('\n') // can't use .lines() or we lose '\n'
                .map(|l| l.chars().collect())
                .collect(),
            line: 0,
            column: 0,
        }
    }
}

impl Iterator for StreamIter<char> {
    type Item = ContextElement<char>;

    fn next(&mut self) -> Option<Self::Item> {
        let opt = self.lines.get(self.line);
        let line = match opt {
            Some(l) => l,
            None => return Some(ContextElement::new(0 as char, self.line, self.column - 1)),
        };
        let cc = line
            .get(self.column)
            .map(|c| ContextElement::new(*c, self.line, self.column))
            .or_else(|| {
                self.line += 1;
                self.column = 0;
                self.lines.get(self.line).and_then(|line| {
                    line.get(self.column)
                        .map(|c| ContextElement::new(*c, self.line, self.column))
                })
            });
        self.column += 1;
        cc.or_else(|| Some(ContextElement::new(0 as char, self.line, self.column - 1)))
    }
}

#[derive(Debug, PartialEq, Serialize)]
pub(crate) struct LexError {
    message: String,
    line: usize,
    column: usize,
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Lexing error: {} at {}:{}",
            self.message, self.line, self.column
        )
    }
}

impl std::error::Error for LexError {}

impl<T> From<(String, ContextElement<T>)> for LexError {
    fn from((msg, cp): (String, ContextElement<T>)) -> Self {
        LexError {
            message: msg,
            line: cp.line,
            column: cp.column,
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_char() {
        let inputs = vec!["'c'", "'\\n'", "'c", "'mm'", "'", "'\\c'", "''"];
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            for input in inputs {
                insta::assert_yaml_snapshot!((input, Lexer::new(input).scan()));
            }
        });
    }

    #[test]
    fn test_cond() {
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            let inputs = vec![
"if x > -3 {
    print(x)
}
",
"if x > -3 {
    print(x)
} else {
    exit()
}
",
"if x > 3 {
    print(x)
} else if y == 1 {
    exit()
} else {
    z
}
"];
            for input in inputs {
                insta::assert_yaml_snapshot!((input, Lexer::new(input).scan()));
            }
        });
    }

    #[test]
    fn test_extern() {
        let input = "extern cos(x)";
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            insta::assert_yaml_snapshot!((input, Lexer::new(input).scan()));
        });
    }

    #[test]
    fn test_fn() {
        let input = "\
fn foo(a, b, c) -> int {
    bar(a) + 2
}";
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            insta::assert_yaml_snapshot!((input, Lexer::new(input).scan()));
        });
    }

    #[test]
    fn test_for() {
        let input = "\
for let x = 1; x < 10; 1 {
    print(x)
}";
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            insta::assert_yaml_snapshot!((input, Lexer::new(input).scan()));
        });
    }

    #[test]
    fn test_logical_ops() {
        let inputs = vec!["x && y", "x || y", "x == y"];
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            for input in inputs {
                insta::assert_yaml_snapshot!((input, Lexer::new(input).scan()));
            }
        });
    }

    #[test]
    fn test_let() {
        let inputs = vec![
            "let a: int = 1",
            "let b: int32 = 2",
            "let b: int64 = 3",
            "let b: uint = 4",
            "let b: uint32 = 5",
            "let b: uint64 = 6",
            "let c: float = 7.0",
            "let c: double = 8.0",
            "let d: bool = true",
            "let d: bool = false",
            "let e: char = 'c'",
            "let e: char = '\n'",
        ];
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            for input in inputs {
                insta::assert_yaml_snapshot!((input, Lexer::new(input).scan()));
            }
        });
    }

    #[test]
    fn test_multi_line_comment() {
        let input = "\
let foo = 14
// line1
// line2
foo
";
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            insta::assert_yaml_snapshot!((input, Lexer::new(input).scan()));
        });
    }

    #[test]
    fn test_ops() {
        let input = "(x + y) * 4 / 4";
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            insta::assert_yaml_snapshot!((input, Lexer::new(input).scan()));
        });
    }

    #[test]
    fn test_semi() {
        let inputs = vec!["a", "1", "a()", "{ a }", "let a: int", "let", "fn", "'c'"];
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            for input in inputs {
                insta::assert_yaml_snapshot!((input, Lexer::new(input).scan()));
            }
        });
    }

    #[test]
    fn test_trailing_comment() {
        let input = "\
let foo = 13
// line2";
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            insta::assert_yaml_snapshot!((input, Lexer::new(input).scan()));
        });
    }

    #[test]
    fn test_unknown_char() {
        let input = "let foo = `1`";
        insta::with_settings!({ snapshot_path => "tests/snapshots", prepend_module_to_snapshot => false }, {
            insta::assert_yaml_snapshot!((input, Lexer::new(input).scan()));
        });
    }

    #[test]
    fn test_lex_one() {
        use Symbol::*;
        use TokenType::*;

        let input = "\
foo + bar
// bar
/ not_a_comment
baz
";
        let mut lexer = Lexer::new(input);

        assert_eq!(lexer.lex(), Ok(Token::new(Ident("foo".to_string()), 1, 1)));
        assert_eq!(lexer.lex(), Ok(Token::new(Op(Add), 1, 5)));
        assert_eq!(lexer.lex(), Ok(Token::new(Ident("bar".to_string()), 1, 7)));
        assert_eq!(lexer.lex(), Ok(Token::new(Op(Div), 3, 1)));
        assert_eq!(
            lexer.lex(),
            Ok(Token::new(Ident("not_a_comment".to_string()), 3, 3))
        );
        assert_eq!(lexer.lex(), Ok(Token::new(Ident("baz".to_string()), 4, 1)));
        assert_eq!(lexer.lex(), Ok(Token::new(Eof, 5, 1)));
    }

    #[test]
    fn test_stream_next() {
        let input = "\
abc
def
ghi
";
        let mut stream = StreamIter::new(input);
        assert_eq!(ContextElement::new('a', 0, 0), stream.next().unwrap());
        assert_eq!(ContextElement::new('b', 0, 1), stream.next().unwrap());
        assert_eq!(ContextElement::new('c', 0, 2), stream.next().unwrap());
        assert_eq!(ContextElement::new('\n', 0, 3), stream.next().unwrap());
        assert_eq!(ContextElement::new('d', 1, 0), stream.next().unwrap());
        assert_eq!(ContextElement::new('e', 1, 1), stream.next().unwrap());
        assert_eq!(ContextElement::new('f', 1, 2), stream.next().unwrap());
        assert_eq!(ContextElement::new('\n', 1, 3), stream.next().unwrap());
        assert_eq!(ContextElement::new('g', 2, 0), stream.next().unwrap());
        assert_eq!(ContextElement::new('h', 2, 1), stream.next().unwrap());
        assert_eq!(ContextElement::new('i', 2, 2), stream.next().unwrap());
        assert_eq!(ContextElement::new('\n', 2, 3), stream.next().unwrap());
        assert_eq!(ContextElement::new(0 as char, 3, 0), stream.next().unwrap());
        assert_eq!(ContextElement::new(0 as char, 3, 0), stream.next().unwrap());
    }
}
