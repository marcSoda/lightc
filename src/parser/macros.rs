#[macro_export]
macro_rules! expect_next_token {
    // Matches patterns like TokenType::Let
    ($ts:expr, $( $e:tt::$v:tt )|+ , $err:expr) => {
        let t = $ts.next();
        match t {
            $( Some(Token { tt: $e::$v, .. }) => (), )+
            _ => {
                return Err(ParseError::from((
                    format!(
                        "{}. Got {}",
                        $err.to_string(),
                        t.map_or(TokenType::Eof, |x| x.tt.clone())
                    ),
                    t.unwrap_or(&Token::default())
                )))
            }
        }
    };

    // Matches patterns like TokenType::Op(Symbol::Assign)
    ($ts:expr, $( $e:tt::$v:tt(Symbol::$s:tt) )|+ , $err:expr) => {
        let t = $ts.next();
        match t {
            $( Some(Token { tt: $e::$v(Symbol::$s), .. }) => (), )+
            _ => {
                return Err(ParseError::from((
                    format!(
                        "{}. Got {}",
                        $err.to_string(),
                        t.map_or(TokenType::Eof, |x| x.tt.clone())
                    ),
                    t.unwrap_or(&Token::default())
                )))
            }
        }
    };

    // Matches patterns like TokenType::Ident(_) and used for return value
    ($ts:expr, $t:tt::$v:tt($_:tt), $err:expr) => {
        {
            let t = $ts.next();
            match t {
                Some(Token { tt: $t::$v(t), .. }) => t,
                _ => {
                    return Err(ParseError::from((
                        format!(
                            "{}. Got {}",
                            $err.to_string(),
                            t.map_or(TokenType::Eof, |x| x.tt.clone())
                        ),
                        t.unwrap_or(&Token::default())
                    )))
                }
            }
        }
    };
}

// Matches token type and executes $and or returns None
#[macro_export]
macro_rules! token_is_and_then {
    ($to:expr, $tt:pat, $and:expr) => {
        match $to {
            Some(Token { tt: $tt, .. }) => Some($and),
            _ => None,
        }
    };
}
