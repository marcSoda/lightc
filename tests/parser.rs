use insta::assert_yaml_snapshot;
use insta::glob;
use std::fs::File;
use std::io::BufRead;
use std::io::BufReader;

use lightc::lexer::Lexer;
use lightc::parser::*;

fn ast_to_string(ast: Result<&[AstNode], &String>) -> String {
    match ast {
        Ok(ast) => ast.iter().map(|x| x.to_string()).collect(),
        Err(err) => err.to_string(),
    }
}

#[test]
fn test_parser() {
    glob!("inputs/parser/*.input", |path| {
        let file = File::open(path).expect("Error reading input file");
        let reader = BufReader::new(file);

        // Each line of the input files is meant to be a separate test
        // case. Treat each as its own AST. Including `ast_string` in the output
        // makes it more readable.
        let asts = reader
            .lines()
            .map(|line| {
                let line = line.expect("Error reading input line");
                let tokens = Lexer::new(&line).collect::<Result<Vec<_>, _>>().unwrap();
                let ast = Parser::new(&tokens).parse();
                let ast_string = ast_to_string(ast.as_deref());
                (ast, ast_string)
            })
            .collect::<Vec<_>>();
        assert_yaml_snapshot!(asts);
    });
}
