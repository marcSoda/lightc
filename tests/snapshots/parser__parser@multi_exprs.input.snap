---
source: tests/parser.rs
assertion_line: 36
expression: asts
input_file: tests/inputs/parser/multi_exprs.input

---
- - Ok:
      - Expr:
          BinOp:
            sym: Pow
            lhs:
              Num:
                value: 19
            rhs:
              BinOp:
                sym: Pow
                lhs:
                  Num:
                    value: 21
                rhs:
                  Num:
                    value: 40
      - Expr:
          BinOp:
            sym: Plus
            lhs:
              BinOp:
                sym: Minus
                lhs:
                  Num:
                    value: 19
                rhs:
                  BinOp:
                    sym: Mult
                    lhs:
                      Num:
                        value: 21
                    rhs:
                      Num:
                        value: 20
            rhs:
              Num:
                value: 40
  - (^ 19 (^ 21 40))(+ (- 19 (* 21 20)) 40)

